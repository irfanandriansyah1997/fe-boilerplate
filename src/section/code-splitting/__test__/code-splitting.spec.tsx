/* eslint-disable import/no-extraneous-dependencies */
import { act, render, waitFor } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Route, Router } from 'react-router-dom';

import '../../../components/organisms/markdown-reader/__test__/markdown-reader-mock.spec';
import CodeSplitting from '..';

/**
 * MOCK_FETCH
 * @returns {Promise<Response>}
 */
const MOCK_FETCH = (): Promise<Response> =>
  new Promise((resolve) => {
    resolve(({
      status: 200,
      text: () => `### Hello World \n Sample Text On Markdown`
    } as unknown) as Response);
  });

describe(`Testing Code Splitting Router`, () => {
  describe(`Snapshot Testing`, () => {
    const tempFetch = global.fetch;

    beforeEach(() => {
      global.fetch = jest.fn().mockImplementation(MOCK_FETCH);
    });

    it(`Testing Routing Part 1, 2 & 3`, async () => {
      const history = createMemoryHistory();

      /**
       * Redirect To Part 1
       */
      act(() => {
        history.push(`/code-splitting/part-1`);
      });

      const { getByTestId, queryByTestId } = render(
        <Router history={history}>
          <Route path="/code-splitting">
            <CodeSplitting />
          </Route>
        </Router>
      );

      await waitFor(() => {
        expect(getByTestId(`code-splitting-1`)).toBeInTheDocument();
        expect(queryByTestId(`code-splitting-2`)).not.toBeInTheDocument();
        expect(queryByTestId(`code-splitting-3`)).not.toBeInTheDocument();

        expect(getByTestId(`code-splitting-1`)).toMatchSnapshot();
      });

      /**
       * Redirect To Part 2
       */
      act(() => {
        history.push(`/code-splitting/part-2`);
      });

      await waitFor(() => {
        expect(queryByTestId(`code-splitting-1`)).not.toBeInTheDocument();
        expect(getByTestId(`code-splitting-2`)).toBeInTheDocument();
        expect(queryByTestId(`code-splitting-3`)).not.toBeInTheDocument();

        expect(getByTestId(`code-splitting-2`)).toMatchSnapshot();
      });

      /**
       * Redirect To Part 3
       */
      act(() => {
        history.push(`/code-splitting/part-3`);
      });

      await waitFor(() => {
        expect(queryByTestId(`code-splitting-1`)).not.toBeInTheDocument();
        expect(queryByTestId(`code-splitting-2`)).not.toBeInTheDocument();
        expect(getByTestId(`code-splitting-3`)).toBeInTheDocument();

        expect(getByTestId(`code-splitting-3`)).toMatchSnapshot();
      });
    });

    afterEach(() => {
      global.fetch = tempFetch;
    });
  });
});
