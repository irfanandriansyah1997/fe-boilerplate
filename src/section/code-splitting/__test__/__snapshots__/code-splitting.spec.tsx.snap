// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Code Splitting Router Snapshot Testing Testing Routing Part 1, 2 & 3 1`] = `
<div
  class="part-1"
  data-testid="code-splitting-1"
>
  <div
    class="navbar-toggle"
    data-testid="navbar-toggle"
  >
    <div
      class="navbar-toggle__text"
    >
      <h6>
        Toggle Custom Component
      </h6>
      <p>
        Set active for showing custom component and inactive for remove custom component
      </p>
    </div>
    <button
      class="navbar-toggle__button"
      data-testid="navbar-toggle-button-active"
      type="button"
    >
      Active
    </button>
    <button
      class="navbar-toggle__button navbar-toggle__button--active"
      data-testid="navbar-toggle-button-inactive"
      type="button"
    >
      Inactive
    </button>
  </div>
</div>
`;

exports[`Testing Code Splitting Router Snapshot Testing Testing Routing Part 1, 2 & 3 2`] = `
<div
  class="part-2"
  data-testid="code-splitting-2"
>
  <div
    class="navbar-toggle"
    data-testid="navbar-toggle"
  >
    <div
      class="navbar-toggle__text"
    >
      <h6>
        Toggle Custom Component
      </h6>
      <p>
        Set active for showing custom component and inactive for remove custom component
      </p>
    </div>
    <button
      class="navbar-toggle__button"
      data-testid="navbar-toggle-button-active"
      type="button"
    >
      Active
    </button>
    <button
      class="navbar-toggle__button navbar-toggle__button--active"
      data-testid="navbar-toggle-button-inactive"
      type="button"
    >
      Inactive
    </button>
  </div>
</div>
`;

exports[`Testing Code Splitting Router Snapshot Testing Testing Routing Part 1, 2 & 3 3`] = `
<div
  class="part-2"
  data-testid="code-splitting-3"
>
  <div
    class="navbar-toggle"
    data-testid="navbar-toggle"
  >
    <div
      class="navbar-toggle__text"
    >
      <h6>
        Toggle Custom Component
      </h6>
      <p>
        Set active for showing custom component and inactive for remove custom component
      </p>
    </div>
    <button
      class="navbar-toggle__button"
      data-testid="navbar-toggle-button-active"
      type="button"
    >
      Active
    </button>
    <button
      class="navbar-toggle__button navbar-toggle__button--active"
      data-testid="navbar-toggle-button-inactive"
      type="button"
    >
      Inactive
    </button>
  </div>
</div>
`;
